### DISCUSSION TITLE
(Regular Expression?) Filter for GUl Console Output

### DISCUSSION THREAD

**Discussion by cerise21** [tbnobody/OpenDTU#1775] [Ideas]
... to limit output lines to those matching an entered regular expression (alternatively: simple string/wildcard match).
Such a feature could facilitate online debugging.

**Comment by MetaChuh** [comment]
hi @cerise21 

there is, imho, not much space on the esp32 left, to implement a good and capable js regex library for that.

recommended for now:
to post process your opendtu console log externally (linux, mac, windows), connect to the opendtu websocket with your favorite grep-tail capable terminal client.

websocket address: `ws://admin:your_password@your_dtu_ip:80/console`
(please replace all `your_*` values with your corresponding values)

for quick tests, you can try this out from any terminal with simple <kbd>curl</kbd>, and also post process it from there, by piping the live output `|` to something else, like <kbd>grep</kbd>

```
curl --include --no-buffer --header "Connection: Upgrade" --header "Upgrade: websocket" --header "Sec-WebSocket-Key: SGVsbG8sIHdvcmxkIQ==" --header "Sec-WebSocket-Version: 13" http://admin:your_password@your_dtu_ip:80/console
```

greetings
metachuh

**Reply by cerise21** [reply]
Yes, I can see the low memory point.
Thanks a lot for describing the web socket approach. I wasn't aware of openDTU providing the protocol.
This should even work using termux/curl/grep on the phone.

**Reply by MetaChuh** [reply]
cool, if you found something, that makes it grepable on a phone, please share 🙏👍

for now, i'm personally using [`websocat`](https://github.com/vi/websocat/releases) together with grep, but not on my phone yet.

note: websocat has portable binaries at the github link above, so no installation is needed, just execute the binary for the desired platform.
(websocat portable binaries are available for: `aarch64-apple-darwin`, `aarch64-linux-android`, `aarch64-linux`, `arm-linux`, `armv7-linux`, `risc-v-linux`, `x86-windows`, `x86_64-apple-darwin`, `x86_64-windows`, `x86_64-freebsd`, etc.)

for example to filter `channel`messages, i use:

`websocat ws://admin:your_password@your_opendtu_ip/console | grep --color=always --ignore-case "Channel"`

greetings

**Reply by cerise21** [reply]
Tested the websocket curl & grep today in the termux shell on my phone (cf. https://termux.dev) - worked.
It is even possible to 
`apt install websocat`
inside termux - great!
The only thing I am missing (using curl or websocat) are the timestamps that are shown at the beginning of each line in openDTU's web console. Are they not part of the websocket's console output?

**Reply by MetaChuh** [reply]
cool thx 🙏

>The only thing I am missing (using curl or websocat) are the timestamps that are shown at the beginning of each line in openDTU's web console. Are they not part of the websocket's console output?

afaik not, they seem to be added at the web interface only, but i only understand appx. 20% of the code yet.
maybe @tbnobody has an answer, if we missed out on something regarding the time stamps.

**Comment by cerise21** [comment]
For completeness:
`websocat -t - timestamp:ws://opendtu_ip/console |grep --line-buffered -i powermeter |tee console.log`
prepends an epoch-style timestamp to every line read from the websocket.
And: replacing `ws://opendtu_ip/console`  by `ws://opendtu_ip/livedata` current Live Data can be retrieved.
